cps/kni.c

```diff
@@ -613,6 +587,152 @@ mnl_socket_recvfrom_flags(const struct mnl_socket *nl, void *buf, size_t bufsiz,
 	return ret;
 }
 
+static int
+new_route(struct route_update *update, struct gk_config *gk_conf)
+{
+	int ret;
+	uint16_t proto;
+	char ip_buf[INET6_ADDRSTRLEN];
+	char gw_buf[INET6_ADDRSTRLEN];
+	char ipp_buf[INET6_ADDRSTRLEN + 4];
+	char gt_buf[INET6_ADDRSTRLEN];
+	int prefix_fib_id, gw_fib_id;
+	struct ipaddr *gt_addr = NULL;
+	struct gk_fib *prefix_fib = NULL;
+	struct gk_fib *gw_fib = NULL;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	if (update->family == AF_INET6) {
+		proto = ETHER_TYPE_IPv6;
+
+		prefix_fib_id = lpm_lookup_ipv6(
+			ltbl->lpm6, update->ip.v6.s6_addr);
+		if (prefix_fib_id < 0) {
+			gw_fib_id = lpm_lookup_ipv6(
+				ltbl->lpm6, update->gw.v6.s6_addr);
+			if (gw_fib_id < 0)
+				return -1;
+			gw_fib = &ltbl->fib_tbl6[gw_fib_id];
+		} else
+			prefix_fib = &ltbl->fib_tbl6[prefix_fib_id];
```

Comment:
I'm a little bit confused about what this code fragment is doing. Could you explain why first we look up `update->ip.v6.s6_addr` and if we don't find an entry then look up `update->gw.v6.s6_addr`?

Reply:
For an FIB entry, it's either needs to direct the ip prefix (i.e., ``update->ip.v6.s6_addr``) to back/front network, or forward it to a grantor server. If it's an ip prefix that needs to be forwarded to a grantor server, then we should have an FIB entry for the prefix. Otherwise, it's possible that we don't have an entry for this prefix, so we need to lookup the FIB entry for the gateway to correctly forward the prefix to the corresponding back/front network.